gems/smithy-client/lib/smithy-client/param_converter.rb

```diff
@@ -0,0 +1,252 @@
+# frozen_string_literal: true
+
+require 'bigdecimal'
+require 'stringio'
+require 'date'
+require 'time'
+require 'tempfile'
+
+module Smithy
+  module Client
+    # @api private
+    class ParamConverter
+      include Smithy::Client::Shapes
+
+      @mutex = Mutex.new
+      @converters = Hash.new { |h, k| h[k] = {} }
+
+      def initialize(rules)
+        @rules = rules
+        @opened_files = []
+      end
+
+      attr_reader :opened_files
+
+      # @param [Hash] params
+      # @return [Hash]
+      def convert(params)
+        structure(@rules, params)
+      end
+
+      def close_opened_files
+        @opened_files.each(&:close)
+        @opened_files = []
+      end
+
+      private
+
+      def structure(shape, values)
+        values = c(shape, values)
+        if values.is_a?(::Struct) || values.is_a?(Hash)
+          values.each_pair do |k, v|
+            next if v.nil?
+            next unless shape.member?(k)
+
+            values[k] = member(shape.member(k), v)
+          end
+        end
+        values
+      end
+
+      def union(shape, values)
```

Comment:
Not at all confident in this - but the intent is to convert all provided unions (maybe something later removes one) but validator will catch if multiple already.

Reply:
I think it makes sense - kick it to validator to catch multiple.